src/box/lua/trigger.c

```diff
@@ -74,20 +135,62 @@ luaT_trigger_call(struct lua_State *L)
 	struct event *event = event_get(event_name, false);
 	if (event == NULL)
 		return 0;
-	int top = lua_gettop(L);
-	int narg = top - 1;
+	int nargs = lua_gettop(L) - 1;
+
+	/*
+	 * If the function is called on Lua-stack from fiber storage, create
+	 * new thread because we cannot dump port_lua to the same stack as
+	 * the underlying one, and it's cheaper to leave fiber storage stack
+	 * to triggers - they won't have to create their own stacks then.
+	 *
+	 * If we have no args, port won't be created, so we don't need to
+	 * create a stack.
+	 */
+	struct lua_State *args_L = L;
+	int args_coro_ref = LUA_REFNIL;
+	if (nargs > 0) {
+		if (fiber()->storage.lua.stack == L) {
+			args_L = luaT_newthread(tarantool_L);
+			if (args_L == NULL)
+				panic("Cannot create Lua thread");
+			args_coro_ref = luaL_ref(tarantool_L,
+						 LUA_REGISTRYINDEX);
+		}
+		lua_xmove(L, args_L, nargs);
+	}
+
+	/*
+	 * Create port_lua starting from the index after current top because
+	 * arguments are popped on call, so we have to push them before each
+	 * trigger.
+	 * If we have no arguments, port is not created.
+	 */
+	struct port *args = NULL;
+	struct port args_port;
+	if (nargs > 0) {
+		port_lua_create_at(&args_port, args_L, lua_gettop(args_L) + 1);
+		args = &args_port;
+	}
+
 	struct event_trigger_iterator it;
 	event_trigger_iterator_create(&it, event);
 	struct func_adapter *trigger = NULL;
 	const char *name = NULL;
 	int rc = 0;
 	while (rc == 0 && event_trigger_iterator_next(&it, &trigger, &name)) {
-		func_adapter_lua_get_func(trigger, L);
-		for (int i = top - narg + 1; i <= top; ++i)
-			lua_pushvalue(L, i);
-		rc = luaT_call(L, narg, 0);
+		/*
+		 * Push arguments before each trigger - they will be popped when
+		 * the trigger is invoked. If nargs is 0, loop won't be started.
+		 */
+		for (int i = 1; i <= nargs; i++)
+			lua_pushvalue(args_L, i);
```

Comment:
Looks like a hack. Better create the port before each call.

Reply:
Decided to move this to `port_lua_dump`, i.e. make it possible to call this function multiple times on the same port.

Reply:
Done as a separate commit.